htlcswitch/circuit.go

```diff
@@ -1,135 +1,169 @@
 package htlcswitch
 
 import (
-	"bytes"
-	"crypto/sha256"
-	"encoding/hex"
+	"fmt"
 	"sync"
 
 	"github.com/go-errors/errors"
 	"github.com/lightningnetwork/lnd/lnwire"
 )
 
-// circuitKey uniquely identifies an active circuit between two open channels.
-// Currently, the payment hash is used to uniquely identify each circuit.
-type circuitKey [sha256.Size]byte
-
-// String returns the string representation of the circuitKey.
-func (k *circuitKey) String() string {
-	return hex.EncodeToString(k[:])
-}
-
-// paymentCircuit is used by the htlc switch subsystem to determine the
-// forwards/backwards path for the settle/fail HTLC messages. A payment circuit
-// will be created once a channel link forwards the htlc add request and
-// removed when we receive settle/fail htlc message.
-type paymentCircuit struct {
+// PaymentCircuit is used by the HTLC switch subsystem to determine the
+// backwards path for the settle/fail HTLC messages. A payment circuit
+// will be created once a channel link forwards the HTLC add request and
+// removed when we receive a settle/fail HTLC message.
+type PaymentCircuit struct {
 	// PaymentHash used as unique identifier of payment.
-	PaymentHash circuitKey
+	PaymentHash [32]byte
 
-	// Src identifies the channel from which add htlc request is came from
+	// SrcChanID identifies the channel from which add htlc request came from
 	// and to which settle/fail htlc request will be returned back. Once
 	// the switch forwards the settle/fail message to the src the circuit
 	// is considered to be completed.
-	Src lnwire.ShortChannelID
+	SrcChanID lnwire.ShortChannelID
+
+	// SrcHTLCID is the ID in the update_add_htlc message we received from the
+	// source channel, which will be included in any settle/fail messages we
+	// send back.
+	SrcHTLCID uint64
 
 	// Dest identifies the channel to which we propagate the htlc add
 	// update and from which we are expecting to receive htlc settle/fail
 	// request back.
-	Dest lnwire.ShortChannelID
+	DestChanID lnwire.ShortChannelID
+
+	// DestHTLCID is the ID in the update_add_htlc message we sent to the
+	// destination source channel.
+	DestHTLCID uint64
 
 	// ErrorEncrypter is used to re-encrypt the onion failure before
 	// sending it back to the originator of the payment.
 	ErrorEncrypter ErrorEncrypter
-
-	// RefCount is used to count the circuits with the same circuit key.
-	RefCount int
 }
 
-// newPaymentCircuit creates new payment circuit instance.
-func newPaymentCircuit(src, dest lnwire.ShortChannelID, key circuitKey,
-	e ErrorEncrypter) *paymentCircuit {
-
-	return &paymentCircuit{
-		Src:            src,
-		Dest:           dest,
-		PaymentHash:    key,
-		RefCount:       1,
-		ErrorEncrypter: e,
-	}
+// circuitKey is a channel ID, hash ID tuple used as an identifying key for a
+// payment circuit.
+type circuitKey struct {
```

Comment:
Not clear from this `godoc` comment if this is the channel id + htlc id or the _incoming_ or _outgoing_ htlc. From reading the diff a bit more, I see it's for the dest. I recommend adding a `dest*` prefix to the fields.